### DISCUSSION TITLE
creating a looming stimulus - issue

### DISCUSSION THREAD

**Discussion by kmelani** [bonsai-rx/bonsai#1429] [Q&A] [answered]
hello! i'm currently an intern working on trying to code a looming stimulus for my project. i'm having a really hard time trying to figure out how to get my stimulus to repeat, as well as how to get the timing from seconds to milliseconds.
to explain: i need a circle to start at my starting diameter, stay there for 250 ms, grow to a larger diameter within 250 ms, stay at this new larger diameter for 500 ms, and then repeat the entire process. the issue is, when it repeats with my coding currently, it doesn't repeat the sequence where the smaller circle stays small before growing. instead, it just repeats the growth of the circle over and over, completely skipping this part of the sequence. 
i have attached a screenshot of my current code. please ignore the disabled part, i've been messing around and trying to figure this out for a while now. 
![bonsai code](https://github.com/bonsai-rx/bonsai/assets/137228932/8c9ae80b-ce89-4cce-a727-55d577d96df4)

**Comment by bruno-f-cruz** [comment]
Can you add the workflow? Thanks

**Reply by kmelani** [reply]
hi! you mean to add the workflow output? where do you suggest i add this? 
i tried to connect it to "draw circle," but it didn't change anything, and i can't connect it to range animation

**Reply by bruno-f-cruz** [reply]
As suggested in #864, include the bonsai file of your workflow.

**Reply by kmelani** [reply]
whenever i try to save it as workflow it doesn’t allow me to for some reason. i got the option once, but when i opened it back up, it was very different and did not work like my original coding. i’m not sure if there’s something i’m missing in order to export workflow? i’m very sorry about this, i’m very new to the coding world and bonsai. i attached a video of what it’s doing with my current code in case that helps at the very least. the timing is set to seconds right now since i dont know how to set time to milliseconds.
i really appreciate your time and assistance. i've been going crazy trying to figure this out. 

https://github.com/bonsai-rx/bonsai/assets/137228932/f27b953c-546a-4319-b685-787e709062c3

**Comment by bruno-f-cruz** [comment]
What happens when you do "save as"?I really won't be able to help much without the code. It also seems to be a priority since even if anyone helps you you won't be able to save the file ...

**Reply by kmelani** [reply]
[looming without timing of circle sizes.zip](https://github.com/bonsai-rx/bonsai/files/11817754/looming.without.timing.of.circle.sizes.zip)

here we go, i think i figured it out. please let me know if this works!!

**Reply by bruno-f-cruz** [reply]
Your workflow does not work there is a "includeworkflow" missing in your zip file

**Reply by kmelani** [reply]
interesting...im not sure what to do about that. i'm restarting my code to see if i can further figure anything out. i'm mostly struggling on finding out the correct way to make my range animation stay the maximum size a certain amount of time before repeating the sequence. do you know if there is a way to time how long the range animation stays the max size?

**Reply by kmelani** [reply]
or how do i get my delay for my animation to repeat? when i connect the "takeuntil" to the drawcircle node, it allows the max size to stay that size for as long as i put it, but when it repeats the delay for the original size before the growth animation starts does not repeat with it. i need to get the delay for the animation to repeat with everything else as well. i'm trying to reattach my new zip file. it's so close to where i need it, the only issue is that when it repeats the original delay for the animation doesn't stay. 
[LOOMING 2 ZIP.zip](https://github.com/bonsai-rx/bonsai/files/11839456/LOOMING.2.ZIP.zip)

**Reply by kmelani** [reply]
also, separately, i keep having an issue with the drawcircle node, to where the diameter changes every time i run the coding. i set it at 0.1 (the size i want the animation to start as) and the end of the animation at 1. whenever i run the code, it runs fine, but then automatically changes the diameter of the drawcircle to the max size of the animation, which messes up the coding for what i want it to do.

**Comment by bruno-f-cruz** [comment, marked as answer]
Would this work?
[LoomingExample.zip](https://github.com/bonsai-rx/bonsai/files/11849051/LoomingExample.zip)

**Reply by kmelani** [reply]
Almost, it seems that the small circle still wasn't staying small for 250 ms before expanding. but I simply added a delay before the rangeanimation to your code, and now it's perfect! thank you so much for being patient and taking the time to help me with this! the only last issue is that whenever i play the code, the "drawcircle" diameter changes and i get this error (in the screenshot). is there any way to fix this error between cycles? or will i just need to edit the starting diameter each time?
![Screenshot 2023-06-23 100201](https://github.com/bonsai-rx/bonsai/assets/137228932/b7325302-680b-48d3-b143-36e731466cc7)

**Reply by bruno-f-cruz** [reply]
I dont understand how can you have that error to be honest... What version of the shaders package and BonVision are you using?

To lock the size for .25seconds use a delay:

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/dd001edc-601f-4c4a-9721-2e87a159aad2)

**Reply by sebus3456789** [reply]
Hi Kmelani,

I'm in the same position as you were back then. How did you solve the problem with the diameter changes?

**Reply by sebus3456789** [reply]
I'm using the shaders library version 0.27.0 published 04.04.2024 and shaders design library version 0.27.0 published 04.04.2024 and the Bonvision version 0.11.0 published 04.04.2024

**Reply by bruno-f-cruz** [reply]
Attach your workflow, and bonsai config file you are trying to use. Thanks!

**Reply by sebus3456789** [reply]
Hey Bruno,

I'm uploading the workflow and the layout file. I basically followed the steps from this video: https://www.youtube.com/watch?v=RTacTdWuwqA&t=1457s&ab_channel=BonVision

On a side note: How do you implement the sweeping so it continously goes from -1 to 1 and back to -1 without it restarting at -1 after reaching 1. If you know what I mean :D  

Thanks for your help!

[Basic_stimulus_3.0.zip](https://github.com/bonsai-rx/bonsai/files/14930563/Basic_stimulus_3.0.zip)

**Reply by bruno-f-cruz** [reply]
The bonsai.config file is located in the bonsai.exe directory. Can you attach it too please? Thanks

**Reply by sebus3456789** [reply]
I hope this file is what you meant, let me know if it isn't. Thanks! 

[Bonsai.zip](https://github.com/bonsai-rx/bonsai/files/14946851/Bonsai.zip)

**Reply by bruno-f-cruz** [reply]
Everything you sent worked the first time on my PC in a clean environment. Not sure how I can help you further.
Make sure you follow this tutorial: https://bonsai-rx.org/docs/articles/environments.html

Regarding your question: "How do you implement the sweeping so it continuously goes from -1 to 1 and back to -1 without it restarting at -1 after reaching 1. If you know what I mean :D", you can concatenate the inverted sequence. Check the refactored workflow attached

[Basic_stimulus_3_refactored.zip](https://github.com/bonsai-rx/bonsai/files/14947476/Basic_stimulus_3_refactored.zip)

**Reply by sebus3456789** [reply]
Okay, that's odd. I will try to delete and reinstall Bonsai. Hopefully it works then. 
I'm going to leave a short recording so you know what it looks like for me.


https://github.com/bonsai-rx/bonsai/assets/126179109/e4f3ad95-85b7-435d-8483-f07e55898381


https://github.com/bonsai-rx/bonsai/assets/126179109/3d64362c-10c0-4a3b-b759-681f36eb59dc

**Reply by bruno-f-cruz** [reply]
You closed the window where draw circle and as trying to be drawn at. I suspect that's the reason why bonsai crashed at the resource no longer exists. Stop bonsai by clicking the stop button instead.
